synapse/cryotank.py

```diff
@@ -571,3 +588,609 @@ def init(self, name, conf=None, timeout=None):
+            self._workq.put((retn, callback, (self, ) + args, kwargs))
+            succ, rv = retn.wait(timeout=self.MAX_WAIT_S)
+            if succ:
+                if isinstance(rv, Exception):
+                    raise rv
+                return rv
+            raise s_exc.TimeOut()
+
+    return wrap
+
+class CryoTankIndexer:
+    '''
+    Manages indexing of a single cryotank's records
+
+    This implements a lazy indexer that indexes a cryotank in a separate thread.
+
+    Cryotank entries are msgpack-encoded json-compatible dictionaries with arbitrary nesting.  An index consists of a
+    property name, one or more datapaths (i.e. what field out of the entry), and a synapse type.   The type specifies
+    the function that normalizes the output of the datapath query into a string or integer.
+
+    Indices can be added and deleted asynchronously from the indexing thread via CryotankIndexer.addIndex and
+    CryotankIndexer.delIndex.
+
+    Indexes can be queried with normValuByPropVal, normRecordsByPropVal, rawRecordsByPropVal.
+
+    To harmonize with LMDB requirements, writing only occurs on the worker thread, while reading indices takes place in
+    the caller's thread.  Both reading and writing index metadata (that is, information about which indices are
+    running) take place on the worker's thread.
+
+    N.B. The indexer cannot detect when a type has changed from underneath itself.   Operators must explicitly delete
+    and re-add the index to avoid mixed normalized data.
+    '''
+    MAX_WAIT_S = 10
+
+    def __init__(self, cryotank):
+        '''
+        Create an indexer.
+
+        Args:
+            cryotank: the cryotank to index
+        Returns:
+            None
+        '''
+        self.cryotank = cryotank
+        ebus = cryotank
+        self._going_down = False
+        self._worker = threading.Thread(target=self._workerloop, name='CryoTankIndexer')
+        path = s_common.gendir(cryotank.path, 'cryo_index.lmdb')
+        cryotank_map_size = cryotank.lmdb.info()['map_size']
+        self._dbenv = lmdb.open(path, writemap=True, metasync=False, max_readers=8, max_dbs=4,
+                                map_size=cryotank_map_size)
+        # iid, v -> offset table
+        self._idxtbl = self._dbenv.open_db(b'indices', dupsort=True)
+        # offset, iid -> normalized prop
+        self._normtbl = self._dbenv.open_db(b'norms')
+        self._to_delete = {}  # type: Dict[str, int]
+        self._workq = s_queue.Queue()
+        # A dict of propname -> version, type, datapath dict
+        self._meta = _IndexMeta(self._dbenv)
+        self._next_offset = self._meta.lowestProgress()
+        self._chunk_sz = 1000  # < How many records to read at a time
+        self._remove_chunk_sz = 1000  # < How many index entries to remove at a time
+        ebus.on('cryotank:puts', self._onData)
+
+        self._worker.start()
+
+        def _onfini():
+            self._going_down = True
+            self._workq.done()
```

Comment:
we should be calling ``fini()`` on this queue object so the ``_workq`` joins asap. otherwise we could end up in a situation where we are waiting for 100000 items to be done (or in progress) and the interpreter gets axed.

Reply:
This work queue isn't going to be 100000 items long. It is going to be a handful of requests from users.